src/jit/gentree.h

```diff
@@ -120,6 +120,8 @@ enum genTreeKinds
     GTK_NOVALUE = 0x0400, // node does not produce a value
     GTK_NOTLIR  = 0x0800, // node is not allowed in LIR
 
+    GTK_NOCONTAIN = 0x1000, // this node is a value, but may not be contained
```

Comment:
> Wouldn't it make more sense to have GTK_CONTAIN and mark only those opers that can actually be contained?

It might, but the list of such nodes is trickier to gather than it appears. `genConsumeRegs` and friends do not generate a complete list; see e.g. https://github.com/dotnet/coreclr/blob/master/src/jit/codegenxarch.cpp#L3363-L3365. There probably aren't too many more situations like this, but we would need to exhaustively enumerate them.

> I'm also more inclined to have this at lower so that a more aggressive and concrete job could be done building address modes before RA.

The process of choosing which nodes to contain should certainly be handled at lowering, but it does seem reasonable to me to hang `canBeContained` off of `CodeGen` instead of `GenTree`.

> Isn't LCL_VAR either memory or reg depending on RA?

Contained `LCL_VAR` nodes are always memory: they are either register candidate lclVars that are currently live on the stack or non-register-candidate lclVars.

> It looks like you could combine IND, LEA, and LCL_VAR into one "contained" mem thing

Each of these cases has subtly different semantics:
- `GT_IND` denotes a memory load from the address defined by its single operand
- `GT_LEA` does not denote a memory access at all--just a (base, index, scale, offset) address mode, where scale and offset are always constants
- `GT_LCL_VAR` when contained denotes a memory load from an implied address

Now, it may be the case that `GT_LEA` most frequently appears as a contained operand to memory loads/stores (`GT_IND`, `GT_STOREIND`, `GT_NULLCHECK`, block ops), so perhaps the semantic difference isn't as much of a barrier as it seems: I can imagine folding `GT_IND` and contained `GT_LCL_VAR` into e.g. `GT_MEM` which would denote a memory access at some address that is described by either a lclVar number or a (possible target-dependent) address mode and obviate the need for `GT_LEA` in these situations. I don't really see this simplifying much unless we add output operands, however, which might be a rather significant design change.